csrc/scheduler/matmul_utils.cpp

```diff
@@ -0,0 +1,794 @@
+  }
+
+  using AxisPos = decltype(std::declval<TensorView>().nDims());
+  const AxisPos unInitPos = -1;
+  AxisPos bcastInApos = unInitPos;
+  AxisPos bcastInBpos = unInitPos;
+
+  // The first and the second input of MMA have the same number of
+  // IterDomains
+  for (AxisPos pos = 0; pos < in_A->nDims(); ++pos) {
+    if (in_A->axis(pos)->isBroadcast()) {
+      if (bcastInApos != unInitPos) {
+        ss << "Mma op first input has more than one broadcast IterDomain: "
+           << std::to_string(bcastInApos) << " and " << std::to_string(pos);
+        return {c10::nullopt, ss.str()};
+      }
+      bcastInApos = pos;
+    }
+    if (in_B->axis(pos)->isBroadcast()) {
+      if (bcastInBpos != unInitPos) {
+        ss << "Mma op second input has more than one broadcast IterDomain: "
+           << std::to_string(bcastInBpos) << " and " << std::to_string(pos);
+        return {c10::nullopt, ss.str()};
+      }
+      bcastInBpos = pos;
+    }
+  }
+
+  // MMA inputs need to have broadcast IterDomains
+  if (bcastInApos == unInitPos || bcastInBpos == unInitPos) {
+    ss << "The " << (bcastInApos == unInitPos ? "first" : "second")
+       << " mma op doesn't have any broadcast IterDomain";
+    return {c10::nullopt, ss.str()};
+  }
+
+  // MMA inputs must have supported data layout, defined in MatmulLayout
+  // MatmulLayout::TT
+  if (bcastInApos == static_cast<size_t>(2) &&
+      bcastInBpos == static_cast<size_t>(0)) {
+    return {MatmulLayout::TT, c10::nullopt};
+  }
+  // MatmulLayout::TN
+  if (bcastInApos == static_cast<size_t>(1) &&
+      bcastInBpos == static_cast<size_t>(0)) {
+    return {MatmulLayout::TN, c10::nullopt};
+  }
+  // MatmulLayout::NT
+  if (bcastInApos == static_cast<size_t>(2) &&
+      bcastInBpos == static_cast<size_t>(1)) {
+    return {MatmulLayout::NT, c10::nullopt};
+  }
+
+  ss << "Unsupported layout, broadcasts: inputA(" << bcastInApos << "), inputB("
+     << bcastInBpos << ")";
+  return {c10::nullopt, ss.str()};
+}
+
+//! A wrapper for core heuristics initialization
+inline bool initCoreHeuristics(
+    std::shared_ptr<MatmulParams> params,
+    const MmaOptions::MacroType& mma_op,
+    const MatmulLayout& layout,
+    const ProblemShape& problem_shape) {
+  const GemmTile instruction_tile = getMmaOpShape(mma_op);
+  GemmTile warp_tile = {-1, -1, -1};
+  GemmTile cta_tile = {-1, -1, -1};
+
+  using DimType = decltype(GemmTile::m);
+
+  // warp tile shape
+  {
+    if (isAmpere(mma_op)) {
+      // Initial target:
+      // - 2 MMA ops per thread in a warp (32 threads), warp tile should be 64x
+      // instruction tile
+      constexpr DimType target_dim = 64;
+
+      const DimType m_ratio = target_dim / instruction_tile.m;
+      const DimType n_ratio = target_dim / instruction_tile.n;
+      const DimType k_ratio = target_dim / (m_ratio * n_ratio);
+
+      warp_tile = {
+          instruction_tile.m * m_ratio,
+          instruction_tile.n * n_ratio,
+          instruction_tile.k * k_ratio};
```

Comment:
Perhaps I haven't named pieces of the algorithm in a clear way, sorry for that.

The general idea is to:

1.  have a warp tile, represented in elements, that will make each thread in warp (32 in total) have 2 mma ops  (it's for perf),
2. `target_dim` should be indeed named as `num_instructions` or `target_instructions_num`, as I targeted 64 instructions per warp,
3. I calculated ratios which will be used to calculate new `warp_tile` (in elements) based on mma op instruction shape,

for example, for mma instruction 161616 when we target 64 instructions per warp (2x instructions per thread) it will look like this:
```
// ratios:
m_ratio = 4 // 64/16
n_ratio = 4 // 64/16
k_ratio = 4 // 64/(4*4)

// warp tile shape
```
{
64, // result of instruction.m, 16, multiplied by 4
64, // result of instruction.n, 16, multiplied by 4
64, // result of instruction.k, 16, multiplied by 4
}
```
The warp tile shape can change due to:
- instruction shape (if we won't use wmma ops, e.g. 16816 on Ampere)
- adjust number of instructions per thread, instead of 2 we can have 1

Does it make sense?